appmail/models.py

```diff
@@ -427,6 +426,10 @@ class Meta:
         get_latest_by = "timestamp"
         verbose_name = "Email message"
         verbose_name_plural = "Email messages sent"
+        indexes = (
+            # Covering index for admin search.
+            models.Index(fields=['to', 'subject']),
```

Comment:
Should this be two separate indexes? 

My understanding is that this will speed up a search for "to=.." or "to=... & subject=..." but not "subject=...". My knowledge isn't great in this area though, so I may be completely off.

Reply:
Also, aren't indices supposed to be unique?

Reply:
@tim-mccurrach That knowledge is correct but in this scenario the search is for `to` & `subject` every single time so we want an index that "covers" ("satisfies") both. There is not a case on the system where we just search for `to` or `subject` because the only search available is the Django admin one which will utilise both regardless of the query.

As for why two doesn't help in this scenario: https://stackoverflow.com/a/39299521/1774000

NB: I did however mean "composite" and not "covering" - covering is something else.

Reply:
> Also, aren't indices supposed to be unique? As in - there must be more than one record with the same to and subject right?

Indexes, don't have to be unique no. You can have an index on any field, even if it's not unique.

Re. my comment above:

Actually, looks like under the hood django-admin combines the fields into a single OR query anyway, so I don't think my comment above is especially relevant. None-the-less, I'm not sure the `subject` part of that index is helping much.

Reply:
>Also, aren't indices supposed to be unique? As in - there must be more than one record with the same to and subject right?

No, the default index you get when using this in Postgres (btree) performs worse as the cardinality of the data shrinks. Cardinality means "number of value possibilities" in this sense. So there is no point adding an index to say a boolean field or a field with three choice values - because there are only two or three options retrospectively and in those scenarios it is most likely better for the query planner to pick a full table scan rather than use an index which isn't going to be much help.

https://distributedsystemsauthority.com/index-efficiency-and-maintenance-postgresql-12-high-performance-guide-part-5-12/#:~:text=Low%20cardinality%20means%20a%20lot,example%2C%20the%20primary%20key%20column.

Reply:
Indexes != Unique indexes.

Reply:
Sorry, I replied before I saw your reply.

The django-admin won't do a `to` & `subject` search though. That's why I'm not sure the `subject` part of this index will help especially. From the link you shared, it seems like if you have separate indexes though the database will just select the best index to use. This is also interesting (from the link):

> As @MondKin mentioned in comments certain queries can actually use several indexes on the same relation. For example, query with OR clauses like a = 123 OR b = 456 (assuming that there are indexes for both columns). In this case postgres would perform bitmap index scans for both indexes, build a union of resulting bitmaps and use it for bitmap heap scan. In certain conditions the same scheme may be used for AND queries but instead of union there would be an intersection.

This would suggest separate indexes might help for some dbs.

Reply:
You're right about the OR! Could indeed be better to have two indexes.

https://devcenter.heroku.com/articles/postgresql-indexes#multi-column-indexes

Unfortunately there's no real way to find out without applying it and seeing; so I'll swap this implementation to use two indexes instead as I think that'll be the bigger overall win here.

Reply:
Done!

Reply:
> Unfortunately there's no real way to find out without applying it and seeing; 

Yes, that's the really big problem with all of this.